### PR TITLE
gh-99254: remove all unused consts from code objects

### PR DESCRIPTION
**Pull request by iritkatriel**
This reduces the total size of unused consts in the top 100 PyPl packages by about 2%:

Before:

Total: 75 errors; 9,946 files; 235,684 code objects; 3,669,436 lines; 31,309,347 opcodes; 31,073,663 opcode pairs; 12,916,440.0 cache_size; 9,198,802.0 cache wasted; 1,858,819 ops quickened; 44,504 prev extended args; **1,509,350 total size of co_consts**; 189,300 number of co_consts

After:

Total: 75 errors; 9,946 files; 235,684 code objects; 3,669,436 lines; 31,307,877 opcodes; 31,072,193 opcode pairs; 12,915,869.0 cache_size; 9,198,231.0 cache wasted; 1,858,819 ops quickened; 43,034 prev extended args; **1,477,889 total size of co_consts**; 189,286 number of co_consts

* Issue: gh-99254

### CODE REVIEW

**Comment by iritkatriel** [comment]
This reduces the total size of unused consts in the top 100 PyPl packages by about 2%:

Before:

Total: 75 errors; 9,946 files; 235,684 code objects; 3,669,436 lines; 31,309,347 opcodes; 31,073,663 opcode pairs; 12,916,440.0 cache_size; 9,198,802.0 cache wasted; 1,858,819 ops quickened; 44,504 prev extended args; **1,509,350 total size of co_consts**; 189,300 number of co_consts

After:

Total: 75 errors; 9,946 files; 235,684 code objects; 3,669,436 lines; 31,307,877 opcodes; 31,072,193 opcode pairs; 12,915,869.0 cache_size; 9,198,231.0 cache wasted; 1,858,819 ops quickened; 43,034 prev extended args; **1,477,889 total size of co_consts**; 189,286 number of co_consts

File: Misc/NEWS.d/next/Core and Builtins/2022-11-08-17-47-10.gh-issue-99254.RSvyFt.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Compile removes all unused consts from code objects (except the first one, which may be a dicstring).
```
Comment (brandtbucher) [inline Misc/NEWS.d/next/Core and Builtins/2022-11-08-17-47-10.gh-issue-99254.RSvyFt.rst]:
```suggestion
The compiler now removes all unused constants from code objects (except the first one, which may be a docstring).
```

File: Python/compile.c [hunk 2]
```diff
@@ -9731,32 +9731,113 @@ optimize_cfg(cfg_builder *g, PyObject *consts, PyObject *const_cache)
...
 }
 
-// Remove trailing unused constants.
+
 static int
-trim_unused_consts(basicblock *entryblock, PyObject *consts)
+remove_unused_consts(basicblock *entryblock, PyObject *consts)
 {
     assert(PyList_CheckExact(consts));
+    Py_ssize_t nconsts = PyList_Size(consts);
```
Comment (kumaraditya303) [inline Python/compile.c]:
```suggestion
    Py_ssize_t nconsts = PyList_GET_SIZE(consts);
```

File: Python/compile.c [hunk 3]
```diff
@@ -9731,32 +9731,113 @@ optimize_cfg(cfg_builder *g, PyObject *consts, PyObject *const_cache)
...
+        err = 0;
+        goto end;
+    }
+
+    /* move all used consts to the beginning of the consts list */
+    for (Py_ssize_t i = 0; i < n_used_consts; i++) {
+        int old_index = index_map[i];
+        assert(i <= old_index);
+        if (i != old_index) {
+            PyObject *value = PyList_GetItem(consts, index_map[i]);
```
Comment (kumaraditya303) [inline Python/compile.c]:
```suggestion
            PyObject *value = PyList_GET_ITEM(consts, index_map[i]);
```

File: Python/compile.c [hunk 4]
```diff
@@ -9731,32 +9731,113 @@ optimize_cfg(cfg_builder *g, PyObject *consts, PyObject *const_cache)
...
+            assert(value != NULL);
+            int res = PyList_SetItem(consts, i, Py_NewRef(value));
+            assert(res == 0);
```
Comment (kumaraditya303) [inline Python/compile.c:9797]:
```suggestion
```

File: Python/compile.c [hunk 5]
```diff
@@ -9731,32 +9731,113 @@ optimize_cfg(cfg_builder *g, PyObject *consts, PyObject *const_cache)
...
+    }
+
+    /* move all used consts to the beginning of the consts list */
+    for (Py_ssize_t i = 0; i < n_used_consts; i++) {
+        int old_index = index_map[i];
+        assert(i <= old_index);
+        if (i != old_index) {
+            PyObject *value = PyList_GetItem(consts, index_map[i]);
+            assert(value != NULL);
+            int res = PyList_SetItem(consts, i, Py_NewRef(value));
```
Comment (kumaraditya303) [inline Python/compile.c:9796]:
```suggestion
            PyList_SET_ITEM(consts, i, Py_NewRef(value));
```

File: Python/compile.c [hunk 6]
```diff
@@ -9731,32 +9731,113 @@ optimize_cfg(cfg_builder *g, PyObject *consts, PyObject *const_cache)
...
+                int index = b->b_instr[i].i_oparg;
+                b->b_instr[i].i_oparg = reverse_index_map[index];
+            }
+        }
+    }
+
+    err = 0;
+end:
+    if (index_map != NULL) {
+        PyMem_Free(index_map);
```
Comment (thatbirdguythatuknownot) [inline Python/compile.c]: Doesn't `PyMem_Free()` do nothing if the value is `NULL`?

Comment (iritkatriel) [reply Python/compile.c]: Probably, I'll check when I sort out the other issues.

File: Python/compile.c [hunk 7]
```diff
@@ -9706,32 +9706,111 @@ optimize_cfg(cfg_builder *g, PyObject *consts, PyObject *const_cache)
...
+
+    /* adjust const indices in the bytecode */
+    reverse_index_map = PyMem_Malloc(nconsts * sizeof(Py_ssize_t));
+    if (reverse_index_map == NULL) {
+        goto end;
+    }
+    for (Py_ssize_t i = 0; i < nconsts; i++) {
+        reverse_index_map[i] = -1;
+    }
+    for (Py_ssize_t i = 0; i < nconsts; i++) {
```
Comment (sweeneyde) [inline Python/compile.c]: Couldn't this be `i < n_used_consts`?

File: Python/compile.c [hunk 8]
```diff
@@ -9706,32 +9706,111 @@ optimize_cfg(cfg_builder *g, PyObject *consts, PyObject *const_cache)
...
+
+    /* condense consts */
+    Py_ssize_t n_used_consts = 0;
+    for (int i = 0; i < nconsts; i++) {
+        if (index_map[i] != -1) {
+            assert(index_map[i] == i);
+            if (n_used_consts != i) {
+                index_map[i] = -1;
+                index_map[n_used_consts] = i;
+            }
```
Comment (sweeneyde) [inline Python/compile.c]: With the `i < n_unused_consts` change below, couldn't this become
```suggestion
            index_map[n_used_consts] = i;
```
? As long as we regard index_map as an array of length `n_unused_consts` after this point.

**Review by sweeneyde — APPROVED** [review]
The code looks good to me. A couple of little things, but they could go either way.

File: Python/compile.c (hunk 7 above)
Comment (iritkatriel) [reply Python/compile.c]: Yes, good idea. 

There are some test failures on windows that I wasn't unable to figure out. Does anything jump out to you? I tried playing around with the int/ Py_ssize_t, but that didn't fix it (though locally it seemed to).

**Comment by sweeneyde** [comment]
Unfortunately, I can reproduce the Windows failures locally.

I get a segfault when `Py_DECREF(0x00000000ffffffff)` happens in some `POP_TOP` instruction. Running `.\python.bat -m test` with `lltrace=1` and commenting out `dump_stack()` gave output that ended with the following:

<details>

```
Resuming frame for 'Enum.__new__' in module 'enum'
0: RESUME 0
2: LOAD_GLOBAL_BUILTIN 1
14: LOAD_FAST 1
16: CALL_NO_KW_TYPE_1 1
26: LOAD_FAST 0
28: IS_OP 0
30: POP_JUMP_IF_FALSE 2
36: NOP
38: LOAD_FAST 0
40: LOAD_ATTR 2
40: LOAD_ATTR 2
60: LOAD_FAST 1
62: BINARY_SUBSCR_DICT
72: RETURN_VALUE

Resuming frame for 'EnumType.__call__' in module 'enum'
42: RETURN_VALUE

Resuming frame.2: INTERPRETER_EXIT
246: RETURN_VALUE

Resuming frame.2: INTERPRETER_EXIT
522: POP_JUMP_IF_FALSE 2
528: LOAD_GLOBAL_BUILTIN 39
540: LOAD_GLOBAL_MODULE 12
552: CALL_NO_KW_LEN 1
562: LOAD_GLOBAL_MODULE 40
574: COMPARE_OP_INT_JUMP 5
714: LOAD_FAST 3
716: LOAD_GLOBAL_MODULE 12
728: LOAD_FAST 2
730: STORE_SUBSCR_DICT
734: LOAD_GLOBAL_BUILTIN 39
746: LOAD_GLOBAL_MODULE 6
758: CALL_NO_KW_LEN 1
768: LOAD_GLOBAL_MODULE 50
780: COMPARE_OP_INT_JUMP 5
920: LOAD_FAST 3
922: LOAD_GLOBAL_MODULE 6
934: LOAD_FAST 2
936: STORE_SUBSCR_DICT
940: LOAD_FAST 3
942: RETURN_VALUE

Resuming frame for 'compile' in module 're'
28: RETURN_VALUE

Resuming frame for '<module>'
262: STORE_NAME 37
264: EXTENDED_ARG 5
266: LOAD_CONST 174
268: LOAD_CONST 24
270: MAKE_FUNCTION 1
272: STORE_NAME 38
274: EXTENDED_ARG 5
276: LOAD_CONST 174
278: LOAD_CONST 25
280: MAKE_FUNCTION 1
282: STORE_NAME 39
284: EXTENDED_ARG 5
286: LOAD_CONST 174
288: LOAD_CONST 26
290: MAKE_FUNCTION 1
292: STORE_NAME 40
294: EXTENDED_ARG 5
296: LOAD_CONST 175
298: LOAD_CONST 27
300: MAKE_FUNCTION 1
302: STORE_NAME 41
304: LOAD_CONST 28
306: MAKE_FUNCTION 0
308: STORE_NAME 7
310: LOAD_CONST 29
312: MAKE_FUNCTION 0
314: STORE_NAME 42
316: EXTENDED_ARG 5
318: LOAD_CONST 174
320: LOAD_CONST 30
322: MAKE_FUNCTION 1
324: STORE_NAME 43
326: LOAD_NAME 44
328: BUILD_TUPLE 1
330: LOAD_CONST 31
332: MAKE_FUNCTION 1
334: STORE_NAME 45
336: LOAD_CONST 32
338: MAKE_FUNCTION 0
340: STORE_NAME 46
342: LOAD_CONST 33
344: MAKE_FUNCTION 0
346: STORE_NAME 47
348: LOAD_NAME 26
350: STORE_NAME 48
352: LOAD_CONST 34
354: MAKE_FUNCTION 0
356: STORE_NAME 49
358: LOAD_CONST 35
360: MAKE_FUNCTION 0
362: STORE_NAME 50
364: LOAD_CONST 36
366: MAKE_FUNCTION 0
368: STORE_NAME 51
370: LOAD_CONST 37
372: MAKE_FUNCTION 0
374: STORE_NAME 52
376: LOAD_CONST 38
378: MAKE_FUNCTION 0
380: STORE_NAME 53
382: EXTENDED_ARG 5
384: LOAD_CONST 176
386: LOAD_CONST 39
388: MAKE_FUNCTION 1
390: STORE_NAME 54
392: LOAD_NAME 18
394: BUILD_TUPLE 1
396: LOAD_CONST 40
398: MAKE_FUNCTION 1
400: STORE_NAME 55
402: EXTENDED_ARG 5
404: LOAD_CONST 172
406: LOAD_CONST 41
408: MAKE_FUNCTION 1
410: STORE_NAME 26
412: LOAD_NAME 16
414: BUILD_TUPLE 1
416: LOAD_CONST 42
418: MAKE_FUNCTION 1
420: STORE_NAME 56
422: NOP
424: LOAD_CONST 1
426: LOAD_CONST 43
428: IMPORT_NAME 13
430: IMPORT_FROM 57
432: STORE_NAME 57
434: POP_TOP
436: NOP
438: LOAD_NAME 58
8846: POP_JUMP_IF_FALSE 9
8848: POP_TOP
```
</details>

Maybe `remove_unused_consts` needs to be moved to earlier in the compilation process so that `EXTENDED_ARG`s won't have been added yet?

**Comment by iritkatriel** [comment]
> Maybe the `remove_unused_consts` call needs to be moved earlier, to the `/** Optimization **/` section, so that `EXTENDED_ARG`s won't have been added yet.

I think it is before - extended args are added in assemble_emit.

**Comment by brandtbucher** [comment]
> I think it is before - extended args are added in assemble_emit.

But jump distances are computed in `assemble_jump_offsets`, which happens before this. If any `EXTENDED_ARG`s are removed from `LOAD_CONST` instructions, those jumps will be wrong.

File: Python/compile.c [hunk 9]
```diff
@@ -8844,7 +8844,7 @@ assemble(struct compiler *c, int addNone)
     /* Can't modify the bytecode after computing jump offsets. */
     assemble_jump_offsets(g->g_entryblock);
 
-    if (trim_unused_consts(g->g_entryblock, consts)) {
+    if (remove_unused_consts(g->g_entryblock, consts)) {
         goto error;
     }
```
Comment (brandtbucher) [inline Python/compile.c]:
```suggestion
    if (remove_unused_consts(g->g_entryblock, consts)) {
        goto error;
    }

    /* Can't modify the bytecode after computing jump offsets. */
    assemble_jump_offsets(g->g_entryblock);
```

**Review by brandtbucher — DISMISSED** [review]

**Comment by iritkatriel** [comment]
> > I think it is before - extended args are added in assemble_emit.
> 
> But jump distances are computed in `assemble_jump_offsets`, which happens before this. If any `EXTENDED_ARG`s are removed from `LOAD_CONST` instructions, anything that jumps over them will be wrong.

Ah right, there's a comment about this in the code.

**Comment by iritkatriel** [comment]
Ok, that fixed it. I wonder why it only showed up on windows.

**Comment by brandtbucher** [comment]
> I wonder why it only showed up on windows.

I bet the startup sequence contains quite a bit of Windows-specific code. Probably just a really lucky/unlucky code path.

**Comment by iritkatriel** [comment]
I have made the requested changes; please review again.

**Comment by iritkatriel** [comment]
Thanks for the reviews and debugging help.

**Pull request merged**